toolchain/check/import.cpp

```diff
@@ -87,10 +72,56 @@ static auto CacheCopiedNamespace(
       << to_scope_id;
 }
 
+// Copies a namespace from the import IR, returning its ID. This may diagnose
+// name conflicts, but that won't change the result because namespaces supersede
+// other names in conflicts.
+static auto CopySingleNameScopeFromImportIR(
+    Context& context,
+    llvm::DenseMap<SemIR::NameScopeId, SemIR::NameScopeId>& copied_namespaces,
+    SemIR::NameScopeId import_scope_id, SemIR::NameScopeId enclosing_scope_id,
+    SemIR::NameId name_id, SemIR::TypeId namespace_type_id)
+    -> SemIR::NameScopeId {
+  auto& scope = context.name_scopes().Get(enclosing_scope_id);
+  auto [it, success] = scope.names.insert({name_id, SemIR::InstId::Invalid});
+  if (!success) {
+    if (auto namespace_inst =
+            context.insts().TryGetAs<SemIR::Namespace>(it->second)) {
+      // Namespaces are open, so we can append to the existing one even if it
+      // comes from a different file.
+      CacheCopiedNamespace(copied_namespaces, import_scope_id,
+                           namespace_inst->name_scope_id);
+      return namespace_inst->name_scope_id;
+    }
+  }
+
+  // Produce the namespace for the entry. Use the invalid node because there's
+  // no node to associate with.
+  auto namespace_inst =
+      SemIR::Namespace{namespace_type_id, name_id, SemIR::NameScopeId::Invalid};
+  auto namespace_id =
+      context.AddPlaceholderInst({Parse::NodeId::Invalid, namespace_inst});
+  namespace_inst.name_scope_id =
+      context.name_scopes().Add(namespace_id, enclosing_scope_id);
+  context.ReplaceInstBeforeConstantUse(
+      namespace_id, {Parse::NodeId::Invalid, namespace_inst});
+
+  // If there's a name conflict, produce the diagnose, but still produce the
+  // namespace to supersede the name conflict in order to avoid repeat
+  // diagnostics.
```

Comment:
Slight wording tweak maybe:

```suggestion
  // Diagnose if there's a name conflict, but still produce the
  // namespace to supersede the name conflict in order to avoid repeat
  // diagnostics.
```